email_notification/admin.py

```diff
@@ -84,7 +90,70 @@ def clean_html_content_sv(self):
 
 
 @admin.register(EmailTemplate)
-class EmailTemplateAdmin(admin.ModelAdmin):
+class EmailTemplateAdmin(ExtraButtonsMixin, admin.ModelAdmin):
     model = EmailTemplate
     form = EmailTemplateAdminForm
     exclude = ["html_content"]
+
+    @button(label="Email Template Testing")
+    def template_tester(self, request):
+        context = self.admin_site.each_context(request)
+        if request.method == "POST":
+            form = EmailTestForm(request.POST)
+            if form.is_valid():
+                context = EmailNotificationContext()
+                context.reservee_name = form.cleaned_data["reservee_name"]
+                context.begin_datetime = form.cleaned_data["begin_datetime"]
+                context.end_datetime = form.cleaned_data["end_datetime"]
+                context.reservation_number = form.cleaned_data["reservation_number"]
+                context.unit_location = form.cleaned_data["unit_location"]
+                context.unit_name = form.cleaned_data["unit_name"]
+                context.reservation_name = form.cleaned_data["reservation_name"]
+                context.reservation_unit_name = form.cleaned_data[
+                    "reservation_unit_name"
+                ]
+                context.price = form.cleaned_data["price"]
+                context.non_subsidised_price = form.cleaned_data["non_subsidised_price"]
+                context.subsidised_price = form.cleaned_data["subsidised_price"]
+                context.tax_percentage = form.cleaned_data["tax_percentage"]
+                context.confirmed_instructions = {
```

Comment:
if you choose to write some class for this logic here, as commented above ☝️ I'd suggest that getting these dicts could have common function.

Reply:
There could be some helper methods to be added in the form class for these 🤔

Reply:
This functionality is now a part of `EmailNotificationContext`. So all we need to do in `admin.py` is:
```
context = EmailNotificationContext.from_form(form)
```

Keeps admin stuff a lot cleaner ☺️